### PR TITLE
feat(expr): add `jsonb_object` function

### PR DESCRIPTION
**Pull request by wangrunji0408**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

As title. 
Due to the bug in parsing array literals, some test cases are not enabled.

## Checklist

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [ ] My PR contains breaking changes. (If it deprecates some features, please create a tracking issue to remove them in the future).
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)
- [ ] My PR changes performance-critical code. (Please run macro/micro-benchmarks and show the results.)

- [ ] My PR contains critical fixes that are necessary to be merged into the latest release. (Please check out the [details](https://github.com/risingwavelabs/risingwave/blob/main/CONTRIBUTING.md))

## Documentation

- [x] My PR needs documentation updates. (Please use the **Release note** section below to summarize the impact on users)

## Release note

Support the following functions:

```
jsonb_object ( text[] ) → jsonb
# Builds a JSON object out of a text array. The array must have exactly one dimension with an even number of members, in which case they are taken as alternating key/value pairs. All values are converted to JSON strings.
Example: jsonb_object('{a, 1, b, def, c, 3.5}') → {"a" : "1", "b" : "def", "c" : "3.5"}

jsonb_object ( text[][] ) → jsonb
# Builds a JSON object out of a text array. The array must have two dimensions such that each inner array has exactly two elements, which are taken as a key/value pair. All values are converted to JSON strings.
Example: jsonb_object('{{a, 1}, {b, def}, {c, 3.5}}') → {"a" : "1", "b" : "def", "c" : "3.5"}

jsonb_object ( keys text[], values text[] ) → jsonb
# This form of jsonb_object takes keys and values pairwise from separate text arrays. Otherwise it is identical to the one-argument form.
Example: jsonb_object('{a,b}', '{1,2}') → {"a": "1", "b": "2"}
```

Reference: https://www.postgresql.org/docs/16/functions-json.html#FUNCTIONS-JSON-CREATION-TABLE

### CODE REVIEW

File: src/expr/impl/src/scalar/cast.rs [hunk 1]
```diff
@@ -177,40 +177,68 @@ pub fn str_to_bytea(elem: &str) -> Result<Box<[u8]>> {
 
 // TODO(nanderstabel): optimize for multidimensional List. Depth can be given as a parameter to this
 // function.
+// FIXME(runji): handle escape characters in double quoted strings.
```
Comment (xiangjinwu) [inline src/expr/impl/src/scalar/cast.rs:180]: What about decouple this fix to its own PR? Here you can input `text[]` with the following syntax:
```
select jsonb_object(array['a', '1', 'b', 'def', 'c', '3.5']);
```

File: src/expr/impl/src/scalar/jsonb_object.rs [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+        });
+    }
+    let mut builder = Builder::<Vec<u8>>::new();
+    builder.begin_object();
+    for value in array.iter() {
+        match value {
+            Some(s) => builder.add_string(s.into_utf8()),
+            None => builder.add_null(),
+        }
+    }
```
Comment (xiangjinwu) [inline src/expr/impl/src/scalar/jsonb_object.rs:53]: Key (elements at even position) cannot be null in PostgreSQL:
```
test=# select jsonb_object(array[null, 'b']);
ERROR:  null value not allowed for object key

test=# select jsonb_object(array['a', null]);
 jsonb_object 
--------------
 {"a": null}
(1 row)
```

**Review by xiangjinwu — APPROVED** [review]
rest lgtm

File: src/expr/impl/src/scalar/jsonb_object.rs (hunk 2 above)
Comment (wangrunji0408) [reply src/expr/impl/src/scalar/jsonb_object.rs:53]: Good catch, thanks!

**Comment by wangrunji0408** [comment]
@xiangjinwu By the way, do we plan to be consistent with postgres on array types? i.e. ignore multi-dimensions in the type declaration. In postgres, both one dimension and two dimension input share the same signature `jsonb_object ( text[] ) → jsonb`.

**Comment by xiangjinwu** [comment]
> @xiangjinwu By the way, do we plan to be consistent with postgres on array types? i.e. ignore multi-dimensions in the type declaration. In postgres, both one dimension and two dimension input share the same signature `jsonb_object ( text[] ) → jsonb`.

https://github.com/risingwavelabs/risingwave/issues/3811#issuecomment-1556656529

No. And in this case, separating `{a,b,c,d,e,f}` from `{{a,b},{c,d},{e,f}}` is clearer.

**Comment by BugenZhao** [comment]
> @wangrunji0408 This function lacks the sqllogictests.

https://github.com/risingwavelabs/risingwave/pull/5246

**Pull request merged**